src/derived.rs

```diff
@@ -594,12 +602,40 @@ where
     DB: Database + 'db,
     Q: QueryFunction<DB>,
 {
-    // FIXME(#24) -- handle parallel case
     fn drop(&mut self) {
         if std::thread::panicking() {
+            // In this case, we had installed a `InProgress` marker but we
+            // panicked before it could be removed. At this point, we
+            // therefore "own" unique access to our slot, so we can just
+            // remove the `InProgress` marker.
+
             let mut map = self.map.write();
-            let _ = map.remove(self.key);
+            let old_value = map.remove(self.key);
+            match old_value {
+                Some(QueryState::InProgress { id, waiting }) => {
+                    assert_eq!(id, self.my_id);
+
+                    let waiting = waiting.into_inner();
+
+                    if !waiting.is_empty() {
+                        // FIXME(#24) -- handle parallel case. In
+                        // particular, we ought to notify those
+                        // waiting on us that a panic occurred (they
+                        // can then propagate the panic themselves; or
+                        // perhaps re-execute?).
+                        panic!("FIXME(#24) -- handle parallel case");
+                    }
+                }
+
+                // If we don't see an `InProgress` marker, something
+                // has gone horribly wrong. This panic will
+                // (unfortunately) abort the process, but recovery is
+                // not possible.
+                _ => panic!("unexpected query state"),
```

Comment:
I think this signifies a bug in salsa, so perhaps a more direct error message would be better:

```
Unexpected panic during query evaluation, aborting the process.

Please report this bug to https://github.com/salsa-rs/salsa/issues.
```